packages/devextreme/js/__internal/grids/grid_core/editing/m_editing.ts

```diff
@@ -803,6 +800,23 @@ class EditingControllerImpl extends modules.ViewController {
     return { insertInfo, key };
   }
 
+  private ensureChangeKeyValue(change: any) {
+    if (isDefined(change.key)) {
+      return;
+    }
+
+    const keyExpr = this._dataController.key();
+    let keyValue;
+    if (change.data && keyExpr && !Array.isArray(keyExpr)) {
+      keyValue = change.data[keyExpr];
+    }
+    if (!isDefined(keyValue)) {
+      keyValue = generateNewRowTempKey();
+    }
+
+    change.key = keyValue;
+  }
```

Comment:
Let's change this method and don't mutate the ```change.key``` property.
And let's refactor it a little :)

Something like this:
```suggestion
private getChangeKeyValue(change: any): unknown {
    if (isDefined(change.key)) {
      return change.key;
    }

    const keyExpr = this._dataController.key();
    const isKeyValueInChangeData = change.data && keyExpr && !Array.isArray(keyExpr);

    return isKeyValueInChangeData
      ? change.data[keyExpr]
      : generateNewRowTempKey();
  }
```

And we can rewrite the ```change.key``` in the ```_addInsertInfo``` method:

```typescript
// With mutation (old variant):
this.ensureChangeKeyValue(change);

// Without mutation (my suggestion):
change.key = this.getChangeKeyValue(change);
```